openpype/scripts/otio_burnin.py

```diff
@@ -263,6 +219,113 @@ def add_timecode(
 
         self._add_burnin(text, align, options, TIMECODE)
 
+    def add_per_frame_text(
+        self,
+        text,
+        align,
+        frame_start,
+        frame_end,
+        listed_keys,
+        options=None
+    ):
+        if not options:
+            options = ffmpeg_burnins.TimeCodeOptions(**self.options_init)
+
+        options = options.copy()
+        if frame_start is None:
+            frame_start = options["frame_offset"]
+
+        # `frame_end` is only for meassurements of text position
+        if frame_end is None:
+            frame_end = options["frame_end"]
+
+        fps = options.get("fps")
+        if not fps:
+            fps = self.frame_rate
+
+        text_for_size = text
+        if CURRENT_FRAME_SPLITTER in text:
+            expr = self._get_current_frame_expression(frame_start, frame_end)
+            if expr is None:
+                expr = MISSING_KEY_VALUE
+                text_for_size = text_for_size.replace(
+                    CURRENT_FRAME_SPLITTER, MISSING_KEY_VALUE)
+            text = text.replace(CURRENT_FRAME_SPLITTER, expr)
+
+        longest_list = max(
+            {len(item["values"]) for item in listed_keys.values()}
+        )
+        new_listed_keys = [{} for _ in range(longest_list)]
+        longest_value_by_key = {}
+        for key, item in listed_keys.items():
+            values = item["values"]
+            last_value = values[-1] if values else ""
+            for _ in range(longest_list - len(values)):
+                values.append(last_value)
+
+            item_keys = list(item["keys"])
+            fill_data = {}
+            sub_value = fill_data
+            last_item_key = item_keys.pop(-1)
+            for item_key in item_keys:
+                sub_value[item_key] = {}
+                sub_value = sub_value[item_key]
+            key_max_len = 0
+            key_max_value = ""
+            for value, new_values in zip(values, new_listed_keys):
+                sub_value[last_item_key] = value
+                try:
+                    value = key.format(**sub_value)
+                except (TypeError, KeyError, ValueError):
+                    value = MISSING_KEY_VALUE
+                new_values[key] = value
+
+                value_len = len(value)
+                if value_len > key_max_len:
+                    key_max_value = value
+                    key_max_len = value_len
+            longest_value_by_key[key] = key_max_value
```

Comment:
I feel this is at least lacking a comment here and there. It appears quite complex but there's no description as to what's going on here.